controllers/nstemplateset/namespaces.go

```diff
@@ -76,13 +76,72 @@ func (r *namespacesManager) ensure(logger logr.Logger, nsTmplSet *toolchainv1alp
 func (r *namespacesManager) ensureNamespace(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet, tierTemplate *tierTemplate, userNamespace *corev1.Namespace) error {
 	logger.Info("ensuring namespace", "namespace", tierTemplate.typeName, "tier", nsTmplSet.Spec.TierName)
 
-	// create namespace before created inner resources because creating the namespace may take some time
+	namespaceNeedsUpdate := false
 	if userNamespace == nil {
+		// userNamespace does not exist, need to create the namespace
+		namespaceNeedsUpdate = true
+		logger.Info("namespace needs to be created")
+	} else {
+		// userNamespace exists, check if the namespace needs to be updated
+		upToDate, err := r.namespaceHasExpectedLabelsFromTemplate(tierTemplate, userNamespace)
+		if err != nil {
+			return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to get namespace object from template for namespace type '%s'", tierTemplate.typeName)
+		}
+		namespaceNeedsUpdate = !upToDate
+		logger.Info("namespace needs to be updated", "namespace", userNamespace.Name)
+	}
+
+	// create namespace before creating inner resources because creating the namespace may take some time
+	if namespaceNeedsUpdate {
 		return r.ensureNamespaceResource(logger, nsTmplSet, tierTemplate)
 	}
 	return r.ensureInnerNamespaceResources(logger, nsTmplSet, tierTemplate, userNamespace)
 }
 
+// namespaceHasExpectedLabelsFromTemplate checks if the namespace has the expected labels from the template object
```

Comment:
could you please just add a comment that it checks only if the namespace has the labels and doesn't check those that could have been removed?

Reply:
Done in 035cdca7d68d85445177585f27f3d28cdadb2dcd